### PR TITLE
Edit Employee

### PR DESCRIPTION
**Pull request by VillyGH**
fix issue #196

### CODE REVIEW

File: src/src/engine/api/APIManager.ts [hunk 1]
```diff
@@ -618,6 +631,29 @@ class APIManager extends Logger {
...
+
+        if (snap && snap.exists()) {
+            return <EmployeeEditDTO>{
+                firstName: snap.data().firstName,
+                lastName: snap.data().lastName,
+                department: snap.data().department,
+                phoneNumber: snap.data().phoneNumber,
+                jobTitles: snap.data().jobTitles,
+                skills: snap.data().skills,
+                role: snap.data().role
```
Comment (DJLawton) [inline src/src/engine/api/APIManager.ts:649]: juste faire un .data()

File: src/src/engine/components/ComponentEditEmployee.tsx [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                        </Form.Control.Feedback>
+                    </Form.Group>
+                    <Form.Group as={Col} md="4">
+                        <Form.Label>Numéro de téléphone</Form.Label>
+                        <Form.Control
+                            name="phoneNumber"
+                            id="phoneNumber"
+                            required
+                            type="tel"
+                            pattern="^(\+?1 ?)?\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$"
```
Comment (DJLawton) [inline src/src/engine/components/ComponentEditEmployee.tsx:83]: use regex constants

File: src/src/engine/components/ComponentDepartmentList.tsx [hunk 3]
```diff
@@ -97,8 +98,14 @@ export class ComponentDepartmentList extends React.Component<DepartmentListProps
...
      */
     readonly onDataChange = async (department: Department): Promise<void> => {
-        if (this.props.onDataChange !== null && this.props.onDataChange) {
-            await this.props.onDataChange(department);
+        if (this.props.onAddDepartment !== null && this.props.onAddDepartment) {
+            await this.props.onAddDepartment(department);
+        }
+    }
+
+    readonly onEditDepartment = async (department: Department): Promise<void> => {
```
Comment (BlobMaster41) [inline src/src/engine/components/ComponentDepartmentList.tsx:106]: #onEditDepartment

File: src/src/engine/components/ComponentDepartmentList.tsx [hunk 4]
```diff
@@ -97,8 +98,14 @@ export class ComponentDepartmentList extends React.Component<DepartmentListProps
      * @throws {Error} If the parent function is not defined
      */
     readonly onDataChange = async (department: Department): Promise<void> => {
```
Comment (BlobMaster41) [inline src/src/engine/components/ComponentDepartmentList.tsx:100]: #onDataChange

File: src/src/engine/components/ComponentEditDepartment.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                    Close
+                </Button>
+                <Button variant="primary" type="submit">
+                    Save Changes
+                </Button>
+            </Modal.Footer>
+        </Modal>
+    }
+
+    private modalVisibility(visible: boolean) {
```
Comment (BlobMaster41) [inline src/src/engine/components/ComponentEditDepartment.tsx:84]: private modalVisibility(visible: boolean): void {

File: src/src/engine/components/ComponentEditDepartment.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    readonly #handleChange = (event: React.ChangeEvent<HTMLFormElement>): void => {
+        const target = event.target;
+        const value = target.type === "checkbox" ? target.checked : target.value;
+        const name = target.id;
+
+        if (!name) {
+            throw new Error("Id is undefined for element in form.");
+        }
+
+        this.setState({
```
Comment (BlobMaster41) [inline src/src/engine/components/ComponentEditDepartment.tsx:131]: ??...{}, ...{ useless


`this.setState({
...this.state,
...{
[name]: value,
}
});`

File: src/src/engine/components/ComponentEditEmployee.tsx [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        const form = event.currentTarget;
+        let isValid = form.checkValidity();
+
+        event.preventDefault();
+        event.stopPropagation();
+
+        let eventTarget: any = event.target;
+        let formData = new FormData(eventTarget);
+        let formDataObj: EmployeeEditDTO = Object.fromEntries(formData.entries()) as unknown as EmployeeEditDTO;
+        console.log(formDataObj)
```
Comment (BlobMaster41) [inline src/src/engine/components/ComponentEditEmployee.tsx:165]: remove this log

File: src/src/engine/components/ComponentEditEmployee.tsx [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        });
+        if (errorType === FormErrorType.NO_ERROR && this.props.employeeId) {
+            let employee: EmployeeEditDTO = {
+                firstName: formDataObj.firstName,
+                lastName: formDataObj.lastName,
+                phoneNumber: formDataObj.phoneNumber,
+                department: formDataObj.department,
+                jobTitles: formDataObj.jobTitles ?? [],
+                skills: formDataObj.skills ?? [], // @ts-ignore
+                role: parseInt(formDataObj.role)
```
Comment (BlobMaster41) [inline src/src/engine/components/ComponentEditEmployee.tsx:183]: ts ignore??

File: src/src/engine/components/NavigationBar.tsx [hunk 9]
```diff
@@ -146,17 +146,13 @@ export class NavigationBar extends React.Component<unknown, NavigationBarState>
      * @see NavigationBarState
      * @private
      */
-    private adminCommandLinks(): JSX.Element {
+    private adminCommandLinks(): JSX.Element | undefined {
```
Comment (BlobMaster41) [inline src/src/engine/components/NavigationBar.tsx:149]: can throw undefined in a render

File: src/src/engine/pages/EditEmployee.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        departments: [],
+        roles: [],
+        titles: [],
+        editedEmployee: undefined
+    }
+
+    public async componentDidMount() : Promise<void> {
+        document.title = "Modifier un Employé - TaskMaster";
+
+        // TODO Add error handling
```
Comment (BlobMaster41) [inline src/src/engine/pages/EditEmployee.tsx:33]: Do it now?

File: src/src/engine/types/Department.ts [hunk 11]
```diff
@@ -10,10 +11,12 @@ export interface DepartmentListState {
  * Contient tous les renseignements des employés de l'application web
  */
 export class Department {
-    name: string;
-    director: string;
+    public departmentId?: string;
```
Comment (BlobMaster41) [inline src/src/engine/types/Department.ts:14]: A department should not have a potential id. It should always have an id. This shoudnt be an option?

**Review by BlobMaster41 — CHANGES_REQUESTED** [review]
Verify

File: src/src/engine/api/APIManager.ts (hunk 1 above)
Comment (BlobMaster41) [reply src/src/engine/api/APIManager.ts:649]: dans une variable et verifier que data est definie

File: src/src/engine/types/Department.ts (hunk 11 above)
Comment (VillyGH) [reply src/src/engine/types/Department.ts:14]: modified next PR

**Review by DJLawton — APPROVED** [review]
looks good

**Pull request merged**